pkg/bgpv1/agent/controller.go

```diff
@@ -221,33 +239,62 @@ func NewController(ctx context.Context, clientset client.Clientset, rtMgr BGPRou
 	}
 
 	// setup policy lister and informer.
-	policyLister := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
-	policyInformer := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
-	policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	c.PolicyLister = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
+	c.policyInformer = factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
+	c.policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    sig.Event,
 		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
 		DeleteFunc: sig.Event,
 	})
-	go policyInformer.Run(stop)
-
-	c := Controller{
-		NodeSpec:     nodeSpecer,
-		PolicyLister: policyLister,
-		Sig:          sig,
-		BGPMgr:       rtMgr,
-	}
 
 	// apply any options.
-	for _, opt := range opts {
+	for _, opt := range params.Opts {
 		opt(&c)
 	}
 
-	// stop channel is passed here, if ctx is canceled the passed stop chan
-	// is closed as well.
-	go c.Run(ctx, stop)
+	params.Lifecycle.Append(&c)
+
 	return &c, nil
 }
 
+// Start is called by hive after all of our dependencies have been started.
+func (c *Controller) Start(_ hive.HookContext) error {
+	c.workerpool = workerpool.New(3)
+	c.workerpool.Submit("policy-informer", func(ctx context.Context) error {
+		c.policyInformer.Run(ctx.Done())
+		return nil
+	})
+
+	c.workerpool.Submit("nodespecer", func(ctx context.Context) error {
+		c.NodeSpec.Run(ctx)
+		return nil
+	})
+
+	c.workerpool.Submit("controller", func(ctx context.Context) error {
+		c.Run(ctx)
+		return nil
+	})
+
+	return nil
+}
+
+// Stop is called by hive upon shutdown, after all of our dependants have been stopped.
+// We should perform a graceful shutdown and return as soon as done or when the stop context is done.
+func (c *Controller) Stop(ctx hive.HookContext) error {
+	doneChan := make(chan struct{})
+	go func() {
+		c.workerpool.Close()
+		close(doneChan)
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
This should return `ctx.Err()` (maybe augmented with "workerpool didn't close in time") as otherwise it hides the fact that `c.workerpool.Close()` took too long and leaves resources running in the background. E.g. tests might continue onto other tests or benchmarks that might be affected by those resources.

In my code I've not bothered to handle it this way. I've just called `c.workerpool.Close()` directly and ignore the hook context as I can't use it to cancel the stop.

Could perhaps consider just dropping the stop context altogether since it's rarely useful. Start context I think is useful as it allows extracting maximum amount information when the start times out rather than just fataling. And again for max information we want the agent to itself abort the start instead of having e.g. k8s kill it due to failing health checks.